src/Features/Core/Portable/EmbeddedLanguages/StackFrame/StackFrameParser.cs

```diff
@@ -0,0 +1,455 @@
+        /// 
+        /// Ex: [|MyClass.MyMethod(string s)|]
+        /// </summary>
+        private ParseResult<StackFrameMethodDeclarationNode> TryParseMethodDeclaration()
+        {
+            var (success, identifierNode) = TryParseNameNode(scanAtTrivia: true);
+            if (!success)
+            {
+                return ParseResult<StackFrameMethodDeclarationNode>.Abort;
+            }
+
+            if (identifierNode is not StackFrameQualifiedNameNode memberAccessExpression)
+            {
+                return ParseResult<StackFrameMethodDeclarationNode>.Abort;
+            }
+
+            var typeArgumentsResult = TryParseTypeArguments();
+            if (!typeArgumentsResult.Success)
+            {
+                return ParseResult<StackFrameMethodDeclarationNode>.Abort;
+            }
+
+            var argumentsResult = ParseMethodParameters();
+            if (!argumentsResult.Success || argumentsResult.Value is null)
+            {
+                return ParseResult<StackFrameMethodDeclarationNode>.Abort;
+            }
+
+            return new(new(memberAccessExpression, typeArgumentsResult.Value, argumentsResult.Value));
+        }
+
+        /// <summary>
+        /// Parses a <see cref="StackFrameNameNode"/> which could either be a <see cref="StackFrameSimpleNameNode"/> or <see cref="StackFrameQualifiedNameNode" />.
+        /// 
+        /// Nodes will be parsed for arity but not generic type arguments.
+        ///
+        /// <code>
+        /// All of the following are valid nodes, where "$$" marks the parsing starting point, and "[|" + "|]" mark the endpoints of the parsed node excluding trivia
+        ///   * [|$$MyNamespace.MyClass.MyMethod|](string s)
+        ///   * MyClass.MyMethod([|$$string|] s)
+        ///   * MyClass.MyMethod([|$$string[]|] s)
+        ///   * [|$$MyClass`1.MyMethod|](string s)
+        ///   * [|$$MyClass.MyMethod|][T](T t)
+        /// </code>
+        /// 
+        /// </summary>
+        private ParseResult<StackFrameNameNode> TryParseNameNode(bool scanAtTrivia)
+        {
+            var currentIdentifer = _lexer.TryScanIdentifier(scanAtTrivia: scanAtTrivia, scanLeadingWhitespace: true, scanTrailingWhitespace: false);
+            if (!currentIdentifer.HasValue)
+            {
+                // If we can't parse an identifier, no need to abort the parser entirely here. 
+                // The caller should determine if an identifier is required or not.
+                return ParseResult<StackFrameNameNode>.Empty;
+            }
+
+            var identifierParseResult = TryScanGenericTypeIdentifier(currentIdentifer.Value);
+            if (!identifierParseResult.Success)
+            {
+                return ParseResult<StackFrameNameNode>.Abort;
+            }
+
+            RoslynDebug.AssertNotNull(identifierParseResult.Value);
+            var lhs = identifierParseResult.Value;
+
+            var parseResult = TryParseQualifiedName(lhs);
```

Comment:
something feels wonky.  why is this both outside the while loop and inside the while loop?